File: DanmakuFrostMaster/BaseTypeDef.cs
﻿using Windows.UI;

namespace Atelier39
{
    public enum DanmakuMode
    {
        Unknown = 0,
        Rolling = 1,
        Bottom = 4,
        Top = 5,
        ReverseRolling = 6,
        Advanced = 7,
        //Code = 8,
        Subtitle = 9
    }

    public enum DanmakuPool
    {
        Normal = 0,
        Subtitle = 1,
        Special = 2
    }

    public enum DanmakuFontSize
    {
        Smallest = 1,
        Smaller = 2,
        Normal = 3,
        Larger = 4,
        Largest = 5
    }

    public enum DanmakuAlignmentMode
    {
        Default = 0,
        LowerLeft = 1,
        LowerCenter = 2,
        LowerRight = 3,
        MiddleLeft = 4,
        MiddleCenter = 5,
        MiddleRight = 6,
        UpperLeft = 7,
        UpperCenter = 8,
        UpperRight = 9
    }

    public class DanmakuItem
    {
        public static float DefaultBaseFontSize = 22;

        /// <summary>
        /// Used to sort danmaku with the same StartMs
        /// </summary>
        public ulong Id;
        public uint StartMs;
        public bool HasBorder;
        public bool HasOutline = true;
        public bool AllowDensityControl = true;
        public bool IsRealtime;
        public float BaseFontSize;
        public float OutlineSize = 2f;
        public string FontFamilyName;
        public string Text;
        public bool? IsBold;
        public DanmakuMode Mode;
        public Color TextColor = Colors.White;
        public Color OutlineColor = Colors.Black;

        public int Weight;
        public string MidHash;

        #region For Advanced mode

        public float StartX;
        public float StartY;
        public float EndX;
        public float EndY;

        public int MarginLeft;
        public int MarginRight;
        public int MarginBottom;
        public DanmakuAlignmentMode AlignmentMode = DanmakuAlignmentMode.Default;
        public DanmakuAlignmentMode AnchorMode = DanmakuAlignmentMode.UpperLeft;

        public byte StartAlpha;
        public byte EndAlpha;

        public ulong DurationMs;
        public ulong TranslationDurationMs;
        public ulong TranslationDelayMs;
        public ulong AlphaDurationMs;
        public ulong AlphaDelayMs;

        /// <summary>
        /// Degree
        /// </summary>
        public float RotateZ;
        /// <summary>
        /// Degree
        /// </summary>
        public float RotateY;

        public bool KeepDefinedFontSize;

        #endregion
    }

    public static class DanmakuDefaultLayerDef
    {
        public const uint DefaultLayerId = 0;
        public const uint RollingLayerId = 0;
        public const uint ReverseRollingLayerId = 1;
        public const uint TopLayerId = 2;
        public const uint BottomLayerId = 3;
        public const uint AdvancedLayerId = 4;
        public const uint SubtitleLayerId = 5;
        public const uint DefaultLayerCount = 6;
    }
}

File: js/src/BaseTypeDef.js
/**
 * DanmakuFrostMaster - JavaScript Version
 * BaseTypeDef.js - Core type definitions and enumerations
 */

'use strict';

const DanmakuMode = Object.freeze({
    Unknown: 0,
    Rolling: 1,
    Bottom: 4,
    Top: 5,
    ReverseRolling: 6,
    Advanced: 7,
    Subtitle: 9
});

const DanmakuPool = Object.freeze({
    Normal: 0,
    Subtitle: 1,
    Special: 2
});

const DanmakuFontSize = Object.freeze({
    Smallest: 1,
    Smaller: 2,
    Normal: 3,
    Larger: 4,
    Largest: 5
});

const DanmakuAlignmentMode = Object.freeze({
    Default: 0,
    LowerLeft: 1,
    LowerCenter: 2,
    LowerRight: 3,
    MiddleLeft: 4,
    MiddleCenter: 5,
    MiddleRight: 6,
    UpperLeft: 7,
    UpperCenter: 8,
    UpperRight: 9
});

class DanmakuItem {
    constructor() {
        // Used to sort danmaku with the same StartMs
        this.id = 0;
        this.startMs = 0;
        this.hasBorder = false;
        this.hasOutline = true;
        this.allowDensityControl = true;
        this.isRealtime = false;
        this.baseFontSize = DanmakuItem.defaultBaseFontSize;
        this.outlineSize = 2;
        this.fontFamilyName = null;
        this.text = '';
        this.isBold = null;
        this.mode = DanmakuMode.Unknown;
        this.textColor = { r: 255, g: 255, b: 255, a: 255 };
        this.outlineColor = { r: 0, g: 0, b: 0, a: 255 };

        this.weight = 0;
        this.midHash = null;

        // For Advanced mode
        this.startX = 0;
        this.startY = 0;
        this.endX = 0;
        this.endY = 0;

        this.marginLeft = 0;
        this.marginRight = 0;
        this.marginBottom = 0;
        this.alignmentMode = DanmakuAlignmentMode.Default;
        this.anchorMode = DanmakuAlignmentMode.UpperLeft;

        this.startAlpha = 255;
        this.endAlpha = 255;

        this.durationMs = 0;
        this.translationDurationMs = 0;
        this.translationDelayMs = 0;
        this.alphaDurationMs = 0;
        this.alphaDelayMs = 0;

        /** Degree */
        this.rotateZ = 0;
        /** Degree */
        this.rotateY = 0;

        this.keepDefinedFontSize = false;
    }
}

DanmakuItem.defaultBaseFontSize = 22;

const DanmakuDefaultLayerDef = Object.freeze({
    DefaultLayerId: 0,
    RollingLayerId: 0,
    ReverseRollingLayerId: 1,
    TopLayerId: 2,
    BottomLayerId: 3,
    AdvancedLayerId: 4,
    SubtitleLayerId: 5,
    DefaultLayerCount: 6
});

if (typeof module !== 'undefined' && module.exports) {
    module.exports = { DanmakuMode, DanmakuPool, DanmakuFontSize, DanmakuAlignmentMode, DanmakuItem, DanmakuDefaultLayerDef };
}
